plugins/configs/api/plugin.go

```diff
@@ -0,0 +1,81 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/configs"
+)
+
+type ConfigAPIPlugin struct {
+	ServiceAddress string               `toml:"service_address"`
+	Storage        config.StoragePlugin `toml:"storage"`
+	tls.ServerConfig
+
+	api    *api
+	cancel context.CancelFunc
+	server *ConfigAPIService
+
+	Log     telegraf.Logger `toml:"-"`
+	plugins []PluginConfig
+}
+
+func (a *ConfigAPIPlugin) GetName() string {
+	return "api"
+}
+
+func (a *ConfigAPIPlugin) Init(ctx context.Context, cfg *config.Config, agent config.AgentController) error {
+	a.api, a.cancel = newAPI(ctx, cfg, agent)
+	if a.Storage == nil {
+		a.Log.Warn("initializing config-api without storage, changes via the api will not be persisted.")
+	} else {
+		if err := a.Storage.Init(); err != nil {
+			return fmt.Errorf("initializing storage: %w", err)
+		}
+
+		if err := a.Storage.Load("config-api", "plugins", &a.plugins); err != nil {
+			return fmt.Errorf("loading plugin state: %w", err)
+		}
+	}
+
+	// start listening for HTTP requests
+	tlsConfig, err := a.TLSConfig()
+	if err != nil {
+		return err
+	}
+	if a.ServiceAddress == "" {
+		a.ServiceAddress = ":7551"
+	}
+	a.server = newConfigAPIService(&http.Server{
+		Addr:      a.ServiceAddress,
+		TLSConfig: tlsConfig,
+	}, a.api, a.Log)
+
+	a.server.Start()
+	return nil
+}
+
+func (a *ConfigAPIPlugin) Close() error {
+	// shut down server
+	// stop accepting new requests
+	// wait until all requests finish
+	a.server.Stop()
+
+	// store state
+	if a.Storage != nil {
+		if err := a.Storage.Save("config-api", "plugins", &a.plugins); err != nil {
```

Comment:
Storage.Load() is called once on startup and Storage.Save() is called once on shutdown. Boltdb seems like overkill in this use case. Saving to JSON would work just as well and have the advantage of being human readable.

I assume you guys have more in mind for boltdb and storage plugins, maybe more robust approach to saving (like saving periodically or after every config change)? Can you share what you have in mind related to this?

Reply:
The storage is a plugin so we aren't ruling out a json storage plugin in the future. Storage either does or will get written to any time a plugin is added or removed. it's more like a db in that you can write individual records instead of rewriting the whole file every time a single record changes.

Reply:
When you say "Storage either does or will get written to any time a plugin is added or removed" I'm not sure we're looking at the same code. Like I said before, Storage.Save() is called once at shutdown. It's not called anywhere else outside of a unit test.

If you have plans for this to work differently in the future, we need you to communicate them somehow. It would be best to have in a design doc, but it at least needs to be commented in the code. Otherwise there's no way to do a meaningful review.